hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeQueueMetrics.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    // Initialize queue for StateContext.incrementalReportQueue,
+    // containerActionQueue, pipelineActionQueue
+    incrementalReportsQueueMap = new HashMap<>();
+    containerActionQueueMap = new HashMap<>();
+    pipelineActionQueueMap = new HashMap<>();
+  }
+
+  @Override
+  public void getMetrics(MetricsCollector collector, boolean b) {
+    MetricsRecordBuilder builder = collector.addRecord(METRICS_SOURCE_NAME);
+
+    Map<SCMCommandProto.Type, Integer> tmpMap =
+        datanodeStateMachine.getContext().getCommandQueueSummary();
+    for (Map.Entry<SCMCommandProto.Type, MetricsInfo> entry:
+        stateContextCommandQueueMap.entrySet()) {
+      builder.addGauge(entry.getValue(),
+          (long) tmpMap.getOrDefault(entry.getKey(), 0));
+    }
+
+    tmpMap = datanodeStateMachine.getCommandDispatcher()
+        .getQueuedCommandCount();
+    for (Map.Entry<SCMCommandProto.Type, MetricsInfo> entry:
+        commandDispatcherQueueMap.entrySet()) {
+      builder.addGauge(entry.getValue(),
+          (long) tmpMap.getOrDefault(entry.getKey(), 0));
+    }
+
+    for (Map.Entry<InetSocketAddress, MetricsInfo> entry:
+        incrementalReportsQueueMap.entrySet()) {
+      builder.addGauge(entry.getValue(),
+          datanodeStateMachine.getContext()
+              .getIncrementalReportQueueSize().getOrDefault(entry.getKey(), 0));
+    }
+    for (Map.Entry<InetSocketAddress, MetricsInfo> entry:
+        containerActionQueueMap.entrySet()) {
+      builder.addGauge(entry.getValue(),
+          datanodeStateMachine.getContext()
+              .getContainerActionQueueSize().getOrDefault(entry.getKey(), 0));
+    }
+    for (Map.Entry<InetSocketAddress, MetricsInfo> entry:
+        pipelineActionQueueMap.entrySet()) {
+      builder.addGauge(entry.getValue(),
+          datanodeStateMachine.getContext().getPipelineActionQueueSize()
+              .getOrDefault(entry.getKey(), 0));
+    }
+  }
+
+  public static synchronized void unRegister() {
+    instance = null;
+    DefaultMetricsSystem.instance().unregisterSource(METRICS_SOURCE_NAME);
+  }
+
+  public void addEndpoint(InetSocketAddress endpoint) {
+    incrementalReportsQueueMap.computeIfAbsent(endpoint,
+        k -> getMetricsInfo(INCREMENTAL_REPORT_QUEUE_PREFIX,
+            CaseFormat.UPPER_UNDERSCORE
+                .to(CaseFormat.UPPER_CAMEL, k.getHostName())));
+    containerActionQueueMap.computeIfAbsent(endpoint,
+        k -> getMetricsInfo(CONTAINER_ACTION_QUEUE_PREFIX,
+            CaseFormat.UPPER_UNDERSCORE
+                .to(CaseFormat.UPPER_CAMEL, k.getHostName())));
+    pipelineActionQueueMap.computeIfAbsent(endpoint,
+        k -> getMetricsInfo(PIPELINE_ACTION_QUEUE_PREFIX,
+            CaseFormat.UPPER_UNDERSCORE
+                .to(CaseFormat.UPPER_CAMEL, k.getHostName())));
+  }
+
+  private MetricsInfo getMetricsInfo(String prefix, String metricName) {
+    String metric = prefix + WordUtils.capitalize(metricName);
```

Comment:
Do you think we should append count / size / total to the end of the metric name? Most counters seem to end in something like this.